paddle/fluid/operators/mkldnn/mul_mkldnn_op.cc

```diff
@@ -0,0 +1,518 @@
+                                           y_num_col_dims, y_dt, ctx);
+
+    auto output_dim = output->dims();
+    if (output_dim.size() != 2) {
+      output->Resize({x_matrix.dims()[0], y_matrix.dims()[1]});
+    }
+
+    if (this->mul_) {
+      UpdateDataPointers(ctx, output, &x_matrix);
+      return *(this->mul_);
+    }
+
+    auto src_desc =
+        this->CreateMemDescriptor(&x_matrix, x_dt, memory::format::nc);
+    this->x_input_ = this->CreateMemory(src_desc, &x_matrix, false /*is_y*/);
+
+    if (std::is_same<K, float>::value) {
+      const auto trans_y_ = TransposeInputY(&y_matrix, y_dt);
+      this->y_input_ = QuantInputY(trans_y_, scale_y);
+    } else if (std::is_same<K, int8_t>::value) {
+      this->y_input_ = this->TransposeInputY(&y_matrix, y_dt);
+    } /* else { something goes wrong } */
+
+    bool force_fp32_output = ctx.Attr<bool>("force_fp32_output");
+    memory::data_type dst_dt;
+    if (force_fp32_output) {
+      dst_dt = framework::ToMKLDNNDataType(
+          framework::DataTypeTrait<float>::DataType);
+    } else {
+      dst_dt = framework::ToMKLDNNDataType(
+          framework::DataTypeTrait<int8_t>::DataType);
+    }
+
+    auto dst_desc =
+        this->CreateMemDescriptor(output, dst_dt, memory::format::any);
+
+    this->mul_ = CreateMulPrimitive(*(this->x_input_), *(this->y_input_),
+                                    dst_desc, output, ctx);
+
+    return *(this->mul_);
+  }
+
+  void UpdateDataPointers(const ExecutionContext &ctx, Tensor *out,
+                          const Tensor *in) {
+    this->x_input_->set_data_handle(const_cast<T *>(in->data<T>()));
+
+    bool force_fp32_output = ctx.Attr<bool>("force_fp32_output");
+    if (force_fp32_output) {
+      this->output_->set_data_handle(out->mutable_data<float>(ctx.GetPlace()));
+    } else {
+      this->output_->set_data_handle(out->mutable_data<int8_t>(ctx.GetPlace()));
+    }
+
+    if (out->format() == memory::format::format_undef) {
+      auto output_format =
+          this->output_->get_primitive_desc().desc().data.format;
+      out->set_format((memory::format)output_format);
+    }
+  }
+
+  memory ReorderWithScale(const memory::desc &src_desc,
+                          const memory::desc &dst_desc, const void *src_data,
+                          const std::vector<float> &scale) {
+    auto mask = scale.size() > 1 ? 1 : 0;
+    mkldnn::primitive_attr attr;
+    attr.set_output_scales(mask, scale);
+
+    auto src_mem =
+        memory({src_desc, this->engine_}, const_cast<void *>(src_data));
+    auto dst_mem = memory({dst_desc, this->engine_});
+
+    auto reorder_pd = mkldnn::reorder::primitive_desc(
+        src_mem.get_primitive_desc(), dst_mem.get_primitive_desc(), attr);
+
+    auto reorder = mkldnn::reorder(reorder_pd, src_mem, dst_mem);
+    stream(stream::kind::eager).submit({reorder}).wait();
+
+    return dst_mem;
+  }
+
+  memory TransposeInputY(const Tensor *input_y, memory::data_type type) {
```

Comment:
This function can be safely removed. As it doesn't make any use of `type` parameter, an equivalent function inherited from `MulPrimitiveFactory` can be used.